File: 06_04_2025_Largest_Divisible_Subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums.sort()
        dp=[1]*len(nums)
        prev=[-1]*len(nums)
        maxi=0
        for i in range(1,len(nums)):
            for j in range(0,i):
                if (nums[i]%nums[j]==0 and dp[i]<dp[j]+1):
                    dp[i]=dp[j]+1
                    prev[i]=j
            if (dp[i]>dp[maxi]):
                maxi=i
        
        ret=[]
        i=maxi
        while (i>=0):
            ret.append(nums[i])
            i=prev[i]
        return ret

        

        

File: 06_04_2025_Largest_Divisible_Subset.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        vector<int> dp(n, 1), prev(n, -1);
        int maxi = 0;

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0 && dp[i] < dp[j] + 1) {
                    dp[i] = dp[j] + 1;
                    prev[i] = j;
                }
            }
            if (dp[i] > dp[maxi]) {
                maxi = i;
            }
        }

        vector<int> result;
        int i = maxi;
        while (i >= 0) {
            result.push_back(nums[i]);
            i = prev[i];
        }

        return result;
    }
};
